test/e2e/e2e_test.go

```diff
@@ -128,10 +135,61 @@ func TestMain(m *testing.M) {
 		testfiles.AddFileSource(testfiles.RootFileSource{Root: framework.TestContext.RepoRoot})
 	}
 
+	if framework.TestContext.PrepullImages {
+		klog.Infof("Pre-pulling images so that they are cached for the tests.")
+		prepullImages()
+	}
+
 	rand.Seed(time.Now().UnixNano())
 	os.Exit(m.Run())
 }
 
 func TestE2E(t *testing.T) {
 	RunE2ETests(t)
 }
+
+func prepullImages() {
+	// These are typically set by ginkg.It, but we're running these beforehand.
+	gomega.RegisterFailHandler(ginkgo.Fail)
+	defer ginkgo.GinkgoRecover()
+
+	f := framework.NewDefaultFramework("image-prepull")
+	f.BeforeEach()
+	defer f.AfterEach()
+	ns := f.Namespace.Name
+	c := f.ClientSet
+
+	prePulledImages := []int{
+		image.Agnhost,
+		image.BusyBox,
+		image.Nginx,
+		image.Httpd,
+	}
+
+	label := map[string]string{"app": "prepull-daemonset"}
+	var imgPullers []*appsv1.DaemonSet
+	for _, img := range prePulledImages {
+		imgName := image.GetE2EImage(img)
+		config := image.GetConfig(img)
+		dsName := fmt.Sprintf("image-puller-%s-%s", config.GetName(), config.GetVersion())
+
+		dsSpec := daemonset.NewDaemonSet(dsName, imgName, label, nil, nil, nil)
+		ds, err := c.AppsV1().DaemonSets(ns).Create(context.TODO(), dsSpec, metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+		imgPullers = append(imgPullers, ds)
+	}
+
+	// this should not be a multiple of 5, because node status updates
+	// every 5 seconds. See https://github.com/kubernetes/kubernetes/pull/14915.
+	dsRetryPeriod := 9 * time.Second
+	dsRetryTimeout := 5 * time.Minute
+
+	for _, imgPuller := range imgPullers {
+		checkDaemonset := func() (bool, error) {
+			return daemonset.CheckPresentOnAllNodes(f, imgPuller)
```

Comment:
Say I'm bringing up a 100 node cluster, this could pass with 2 nodes up, because the other 98 haven't yet become schedulable.  So maybe you want to wait to run this until you have the number of schedulable nodes you expect.